cmd/nodelink-controller/main.go

```diff
@@ -393,8 +393,22 @@ func (c *Controller) processNode(node *corev1.Node) error {
 		modNode.Labels[k] = v
 	}
 
-	// Taints are to be an authoritative list on the machine spec per cluster-api comments:
-	modNode.Spec.Taints = matchingMachine.Spec.Taints
+	// Taints are to be an authoritative list on the machine spec per cluster-api comments.
+	// However, we believe many components can directly taint a node and there is no direct source of truth that should enforce a single writer of taints
+	for _, mTaint := range matchingMachine.Spec.Taints {
+		glog.V(3).Infof("machine taint: %v", mTaint)
+		alreadyPresent := false
+		for _, nTaint := range modNode.Spec.Taints {
+			if nTaint.Key == mTaint.Key && nTaint.Effect == mTaint.Effect {
+				glog.V(3).Infof("Skipping to add machine taint, %v, to the node. Node already has a taint with same key and effect", mTaint)
+				alreadyPresent = true
+				break
+			}
+		}
+		if !alreadyPresent {
+			modNode.Spec.Taints = append(modNode.Spec.Taints, mTaint)
```

Comment:
we are appending here when `nTaint.Key == mTaint.Key && nTaint.Effect != mTaint.Effect`, what will be the behaviour?

Reply:
This is list of taints and for a taint to be unique key and effect as a pair must be unique. PTAL at example here:
https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/#concepts